pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,296 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedrouter
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+// PatchEnvelopes is a structure representing
+// Patch Envelopes exposed to App instances via metadata server
+// for more info check docs/PATCH-ENVELOPES.md
+// Must be created by calling NewPatchEnvelopes()
+//
+// Internally, PatchEnvelopes structure stores envelopes which
+// come from EdgeDevConfig parsed by zedagent. This envelopes contains
+// both inline binary artifacts which are ready to be downloaded by app instances
+// and volume references, which are handled by volumemgr.
+// So PatchEnvelopes struct has completedVolumes and contentTreeStatus to store
+// information of all volumes and contentTree handled by volumemgr to link them with
+// patch envelope volume references. ContentTreeStatus is used to retrieve SHA of underlying
+// file.
+// PatchEnvelopes are used by app instances via Get() which returns list of available PatchEnvelopeInfo
+// to certain App Instance which are stored in currentState.
+// PatchEnvelopes also has two channels: updateStateNotificationCh and deleteStateNotificationCh
+// to receive notification about the need of updating specified PatchEnvelopes.
+// NewPatchEnvelopes() starts goroutine processStateUpdate() which reads from both channels and updates currentState
+// to desired one.
+// This way handlers can do work of determining which patch envelopes actually need change (if any)
+// and send back in go routine rest of the update including slow work.
+// Note that this channels are only accessible from the outside by calling a function which returns
+// write-only channel, meaning that updateStateNotificationCh and deleteStateNotificationCh should not be
+// read from anywhere except processStateUpdate() so that there could not be any deadlock.
+type PatchEnvelopes struct {
+	sync.Mutex
+
+	updateStateNotificationCh chan []uuid.UUID
+	deleteStateNotificationCh chan []uuid.UUID
```

Comment:
I think a single channel of type chan struct{} 
should be sufficient for notifications is you encode the change/delete needed in the PatchEnvelopeInfo

Reply:
Make two maps, tested with controller